core_api/src/runnables.py

```diff
@@ -10,24 +10,47 @@
     chain,
 )
 from langchain_core.vectorstores import VectorStoreRetriever
+from tiktoken import Encoding
 
 from core_api.src.format import format_chunks
 from redbox.models import ChatResponse
 from redbox.models.chat import SourceDocument
+from redbox.models.errors import AIError
 
 
-def make_chat_prompt_from_messages_runnable(system_prompt: str, question_prompt: str):
+def make_chat_prompt_from_messages_runnable(
+    system_prompt: str,
+    question_prompt: str,
+    context_window_size: int,
+    tokeniser: Encoding,
+):
     system_prompt_message = [("system", system_prompt)]
+    prompts_budget = len(tokeniser.encode(system_prompt)) - len(tokeniser.encode(question_prompt))
+    token_budget = context_window_size - prompts_budget
 
     @chain
     def chat_prompt_from_messages(input_dict: dict):
         """
         Create a ChatPrompTemplate as part of a chain using 'chat_history'.
         Returns the PromptValue using values in the input_dict
         """
+        context_budget = token_budget - len(tokeniser.encode(input_dict["question"]))
+
+        if context_budget <= 0:
+            message = "Question length exceeds context window."
+            raise AIError(message)
```

Comment:
This gives a 500 in the API response? What about Django?

Reply:
Sorry, not sure I understand the Django question.